### PR TITLE
Add NetworkManager dnsmasq configuration dispatcher

### PR DESCRIPTION
**Pull request by sdodson**
Configures dnsmasq for cluster dns and configures /etc/resolv.conf to use
dnsmasq

### CODE REVIEW

**Comment by sdodson** [comment]
@dcbw This seems to work sufficiently, mind taking a look? We'd use ansible or something to ensure that the proper cluster dns and kube svc ip for non default environments

File: contrib/dnsmasq/99-origin-dns.sh [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+      cat << EOF > /etc/dnsmasq.d/origin-dns.conf
+strict-order
+no-resolv
+domain-needed
+server=/cluster.local/172.30.0.1
+server=/30.172.in-addr.arpa/172.30.0.1
+EOF
+    fi
+    # zero out our upstream servers list and feed it into dnsmasq
+    echo '' > /etc/dnsmasq.d/origin-upstream-dns.conf
```
Comment (ncdc) [inline contrib/dnsmasq/99-origin-dns.sh]: minor nit: use `echo -n` to avoid the initial blank line?

Comment (sdodson) [inline contrib/dnsmasq/99-origin-dns.sh]: Good idea, thanks.

File: contrib/dnsmasq/99-origin-dns.sh [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+. ./network-functions
+
+[ -f ../network ] && . ../network
+
+if [[ $2 =~ ^(up|dhcp4-change)$ ]]; then
+  # couldn't find an existing method to determine if the interface owns the 
+  # default route
+  def_route=$(/sbin/ip route list match 0.0.0.0/0 | awk '{print $3 }')
+  def_route_int=$(/sbin/ip route get to ${def_route} | awk '{print $3}')
+  def_route_ip=$(/sbin/ip route get to ${def_route} | awk '{print $5}')
```
Comment (dcbw) [inline contrib/dnsmasq/99-origin-dns.sh]: You can use the IP4_ADDRESS_0 env variable here, which has the format "<ip>/<prefix> <gateway>" if that's easier.

File: contrib/dnsmasq/99-origin-dns.sh [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+cd /etc/sysconfig/network-scripts
+. ./network-functions
+
+[ -f ../network ] && . ../network
+
+if [[ $2 =~ ^(up|dhcp4-change)$ ]]; then
+  # couldn't find an existing method to determine if the interface owns the 
+  # default route
+  def_route=$(/sbin/ip route list match 0.0.0.0/0 | awk '{print $3 }')
```
Comment (dcbw) [inline contrib/dnsmasq/99-origin-dns.sh]: Note that NM allows multiple default routes if multiple interfaces have a connection.  Lowest metric wins.

**Comment by dcbw** [comment]
Looks pretty sane :)

File: contrib/dnsmasq/99-origin-dns.sh (hunk 3 above)
Comment (sdodson) [inline contrib/dnsmasq/99-origin-dns.sh]: I'm not terribly worried about which interface we pick so long as it's an IP address that a pod can reach. A question I had in general is whether or not multitenant networking breaks the ability for pods to hit the node's IP?

Comment (dcbw) [inline contrib/dnsmasq/99-origin-dns.sh]: Nope, should be fine.  At the moment all pods can access the outside world, what's restricted is traffic between/within ClusterNetworkCIDR and ServiceNetworkCIDR.  pod <-> host is OK.

**Comment by sdodson** [comment]
@smarterclayton merge? This just a contrib script people can drop into place if they want to use dnsmasq without having used the installer. This should work under default configurations and a more complete experience that's customized per cluster config will be available in the installer.

**Comment by smarterclayton** [comment]
[merge]

**Comment by sdodson** [comment]
@smarterclayton re [merge] no way a crontib file is breaking tests.

File: contrib/dnsmasq/99-origin-dns.sh [hunk 4]
```diff
@@ -45,7 +45,7 @@ EOF
     fi
     # zero out our upstream servers list and feed it into dnsmasq
     echo -n > /etc/dnsmasq.d/origin-upstream-dns.conf
-    for ns in ${DHCP4_DOMAIN_NAME_SERVERS}; do
+    for ns in ${IP4_NAMESERVERS}; do
```
Comment (sdodson) [inline contrib/dnsmasq/99-origin-dns.sh]: @dcbw decent solution for staticly configured interfaces?

**Comment by sdodson** [comment]
Updated with recent changes made in ansible. @brenton can i get another merge?

**Comment by brenton** [comment]
[merge]

**Pull request merged**